### ISSUE TITLE
Reduce test fixtures size to speed up tests

### ISSUE THREAD

**Issue by ixje** [CityOfZion/neo-python#478]
### Current behavior
travis-ci takes > 15 minutes to run all tests. One reason is that the tests rely on a fixture database that's > 500MB in size. The fixture database is basically a zipped up testnet blockchain folder. This big size is the result of having testcases deployed on the regular TestNet, which then have many unnecessary blocks added to the fixtures in between different tests created on that network.

We believe that the fixture size can be easily reduced to < 50 MB by recreating all tests on a special private TestNet. This testnet should be based on [neo-privatenet-docker](https://github.com/CityOfZion/neo-privatenet-docker)
The image should only be spun up when adding new tests that requires events or interactions on a full blockchain (e.g. smart contract storage tests) to keep the database to the minimal size necessary.

TODO:
- [x] create an inventory of all tests to be ported to the private testnet (hint: look at what derives from: [BlockchainFixtureTestCase](https://github.com/CityOfZion/neo-python/blob/master/neo/Utils/BlockchainFixtureTestCase.
- [x] migrate all tests to a new private testnet docker image (based on neo-privatenet-docker)
- [x] make testnet docker image publicly available
- [x] create new fixture file, get it on CoZ AWS (ask @localhuman), update all test cases.
- [x] create PR to `neo-python`

**Comment by dauTT** [comment]
Hi ixje, I would be happy to help you with this issue. Shall I start with the creation of the inventory? Question: how exactly do you want to migrate all tests using docker? A concrete example would help here as I have basic knowledge of docker. Thanks.

**Comment by ixje** [comment]
Hi @dauTT that would be great!

To explain a bit more; almost all of the current tests are created on the official TestNet. Which means that there are smart contracts deployed on the official TestNet that have been interacted with in the past and we then confirm with our tests that we still get the same results. 

We want to re-create those smart contracts (as the sources are likely lost @localhuman ?), but then deploy them on a new private network created with the `neo-private-docker` image. Have a look at https://github.com/CityOfZion/neo-privatenet-docker to see what that is about.

* So the first step would be to try re-create the smart contracts belonging to the specific tests that are currently in `neo-python`. You can test and validate your new contracts either on the testnet or using a `neo-private-docker`.
* once that is done we create a new private network where we deploy all smart contracts needed for the tests, do the required invocations (=interactions with the contracts) and then create a snapshot of the docker image to keep it small (because if you keep it running, it will keep generating blocks in the background and the blockchain will then unnecessarily increase in size).

I hope that makes sense. Let me know if I can clarify anything else.

**Comment by dauTT** [comment]
Thanks ixje for the clarification. I believe I understand now what to do. I will proceed with migrating only one test (e.g,  neo-python/neo/SmartContract/tests/test_smart_contract2.py) and then after cross checking with you again I will migrate all the remaining test cases.

**Comment by ixje** [comment]
I'll be away for 4 days so I can't give you any support. If you want you can reach out to others like metachris and localhuman on Discord: https://discordapp.com/invite/b8QNXwD/

**Comment by dauTT** [comment]
Hi ixje, 

as @localhuman mentioned on Discord, he doesn't have the source code of the smart contracts used in the neo-python BlockchainFixture test cases. So he suggested that it may be a good idea to modify the test cases and use any smart contract in https://github.com/CityOfZion/neo-boa/tree/master/boa_test/example. I did that for a test case and was able to migrated to the private `neo-private-docker`. I will share the result of my test with you most likely tomorrow.

**Comment by jseagrave21** [comment]
@dauTT Hello, I am wondering if you would like some help with this issue. Not sure how to collaborate in GitHub (I am new to GH) but I'd love to if that is possible.

**Comment by ixje** [comment]
sure, just drop me a message on Discord (same name as here)

**Comment by dauTT** [comment]
@jseagrave21 , you most welcome to join. We will defintely need help here. :)

@ixje : I have create an image of the neo-privnet on the docker.hub:

```docker pull  dautt/neo-privnet-unittest:v0.0.0```

On this privnet I have deploy the smart contract (sample2.py) as described by localhum in his [guideline](https://neo-python.readthedocs.io/en/latest/neo/SmartContract/smartcontracts.html). You can search for the contract by using this command:

`contract search test`

I have tested the image with neo-python `cc4da934d3afc2ad46e1deb98ce638f0afd157bc (tag: v0.7.3)`. If we take that particular version and replace the 2 files:
1. https://github.com/CityOfZion/neo-python/blob/master/neo/SmartContract/tests/test_smart_contract.py
2. https://github.com/CityOfZion/neo-python/blob/master/neo/Utils/BlockchainFixtureTestCase.py

with the corresponding one in my fork:
1.  https://github.com/dauTT/neo-python/blob/move-unittest-to-privnet-478/neo/SmartContract/tests/test_smart_contract.py
2. https://github.com/dauTT/neo-python/blob/move-unittest-to-privnet-478/neo/Utils/BlockchainFixtureTestCase.py

we see that the test cases in  `test_smart_contract.py` for the `block 9369` will pass by running the following commands:

1) run the privnet:
`docker run --rm -d --name neo-privnet-unittest  -p 20333-20336:20333-20336/tcp -p 30333-30336:30333-30336/tcp dautt/neo-privnet-unittest:v0.0.0`
2) activate the virtual environment
3) python prompt.py -p
4) After the local blockchain is in sync with the privnet, exit from the prompt.
5) python3 -m unittest /neo/SmartContract/tests/test_smart_contract.py

Currently the latest neo-python version and the neo-privnet image are not compatible. So my branch move-unittest-to-privnet-478 is also not working properly. Do you know the reason for this discrepancy?

I will try to fix this issue as well, but meanwhile you can already review whatever has been done so far and let us know if we are in right direction. Thanks.

(Below, in the attachment, we find more information about the test case:)
[POC test case.txt](https://github.com/CityOfZion/neo-python/files/2247589/POC.test.case.txt)

**Comment by jseagrave21** [comment]
@dauTT Could you specify exactly what you mean by "the latest neo-python version and the neo-privnet image are not compatible"? What errors/failures/warnings are you receiving?

**Comment by jseagrave21** [comment]
After talking with @ixje, I think I understand how to proceed. I created a branch for this project but we will have to document our changes on this issue because there is no way we can think of to all edit the same code at the same time.
We will need to verify every test manually on the @dauTT's privnet before we can create the final snapshot.
I started to work on the next test: test_smart_contract2.py located here: https://github.com/jseagrave21/neo-python/blob/16133d5236f9e4c7218bf6a5e56480d16eaf8762/neo/SmartContract/tests/test_smart_contract2.py
Unfortunately, when I ran it I get this error:
```
ERROR: test_b_invocation (neo.SmartContract.tests.test_smart_contract2.SmartContractTest2)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/c/users/jseag/Neo/neo-python-travis/neo/SmartContract/tests/test_smart_contract2.py", line 26, in test_b_invocation
    result = self._blockchain.Persist(block)
  File "/mnt/c/users/jseag/Neo/neo-python-travis/neo/Implementations/Blockchains/LevelDB/TestLevelDBBlockchain.py", line 71, in Persist
    if prevTx.outputs[input.PrevIndex].AssetId.ToBytes() == Blockchain.SystemShare().Hash.ToBytes():
AttributeError: 'NoneType' object has no attribute 'outputs'
```
I am not sure why I am receiving this error.
I will keep working.

I am also trying to think about a way to keep track of tests we have "updated". Maybe a Google Doc? What do you guys think? @ixje @dauTT

**Comment by ixje** [comment]
@dauTT looks good what you have so far! At the end of the road we'll have to keep most of your deletions from `BlockchainFixtureTestCase.py`, but for now it's a good approach for testing.

I followed all your instructions (nice a clear 👍) but I do not face any errors

 ```
(venv) Eriks-Air:neo-python erik$ python -m unittest neo/SmartContract/tests/test_smart_contract.py
[I 180801 20:39:28 LevelDBBlockchain:115] Created Blockchain DB at /Users/erik/.neopython/Chains/privnet
...
----------------------------------------------------------------------
Ran 3 tests in 0.408s

OK
```

The only difference is that I took the latest master `0.7.5` and applied your edits, where you run `0.7.3`.

**Comment by ixje** [comment]
A document could work, or you post whatever test file you're going to work on next here. It's probably best to discuss it on Discord or alike though. Looking forward to your contributions 👍

**Comment by dauTT** [comment]
I guess we can proceed to migrate the remaining test cases:

0. [test_LevelDBBlockchain.py](https://github.com/CityOfZion/neo-python/tree/master/neo/Implementations/Blockchains/LevelDB/test_LevelDBBlockchain.py)
1. [test_leveldb.py](https://github.com/CityOfZion/neo-python/tree/master/neo/Implementations/Blockchains/LevelDB/tests/test_leveldb.py)
2. [test_interop_blockchain.py](https://github.com/CityOfZion/neo-python/tree/master/neo/VM/tests/test_interop_blockchain.py)
3. [test_rest_api.py](https://github.com/CityOfZion/neo-python/tree/master/neo/api/REST/test_rest_api.py)
4. [test_json_invoke_rpc_api.py](https://github.com/CityOfZion/neo-python/tree/master/neo/api/JSONRPC/test_json_invoke_rpc_api.py)
5. [test_json_rpc_api.py](https://github.com/CityOfZion/neo-python/tree/master/neo/api/JSONRPC/test_json_rpc_api.py)
6. [test_smart_contract.py](https://github.com/CityOfZion/neo-python/tree/master/neo/SmartContract/tests/test_smart_contract.py) (closed)
7. [test_smart_contract3.py](https://github.com/CityOfZion/neo-python/tree/master/neo/SmartContract/tests/test_smart_contract3.py)
8. [test_smart_contract2.py](https://github.com/CityOfZion/neo-python/tree/master/neo/SmartContract/tests/test_smart_contract2.py) (@jseagrave21  is working on it)

There are not so many. So if you don't mind I can track them here. Please let me know if I have missed anything.

If we need to deploy more contracts on the privnet docker let me know. I can do it for you.

**Comment by dauTT** [comment]
Hi @jseagrave21 , are you interested on working on the following test modules as well?

4. [test_json_invoke_rpc_api.py](https://github.com/CityOfZion/neo-python/tree/master/neo/api/JSONRPC/test_json_invoke_rpc_api.py)
5. [test_json_rpc_api.py](https://github.com/CityOfZion/neo-python/tree/master/neo/api/JSONRPC/test_json_rpc_api.py)
7. [test_smart_contract3.py](https://github.com/CityOfZion/neo-python/tree/master/neo/SmartContract/tests/test_smart_contract3.py)

**Comment by jseagrave21** [comment]
Sure thing. I am stuck on test_smart_contract2.py. Would you mind taking a look at it? I will temporarily move on to test_smart_contract3.py

**Comment by dauTT** [comment]
Okay, I will look at it maybe tomorrow. Thanks! :)

**Comment by jseagrave21** [comment]
@dauTT I completed the update for test_smart_contract2.py
https://github.com/jseagrave21/neo-python/blob/patch-5/neo/SmartContract/tests/test_smart_contract2.py

**Comment by jseagrave21** [comment]
@dauTT For test_smart_contract3, I think it is referencing a block which deploys a smart contract. I asked @ixje if there was a way to search the blocks in the privatenet for one which deploys a SC and he couldn't think of one. I am wondering if you think this test will require us to write a test SC and deploy it on the privatenet then test it with an updated version of test_smart_conract3?

**Comment by dauTT** [comment]
Hi @jseagrave21 ,

I have already deploy a smart contract on privnet. It is related to `block 9369`. For more information you can look at the attachment that I have posted above (POC test case.txt). With respect to the a new SM contract deployment, If it make your life easy, just do it on your privnet for testing. But I will first try to work with what we already have. If you decide to use a new smart contract, remember that once I will deploy on the privnet to create the new image, your test cases won't work anymore and you will need to fix them again. So Unfortunately we need to do double work but if you have figure out how to fix them once, then the second time is quicker.

**Comment by jseagrave21** [comment]
Oh that is great! I will see if I can work with that.

**Comment by jseagrave21** [comment]
@dauTT @ixje okay test_smart_contract3 is done I think. Please review to make sure it is still doing what we want it to do.

https://github.com/jseagrave21/neo-python/blob/patch-5/neo/SmartContract/tests/test_smart_contract3.py

**Comment by dauTT** [comment]
Still working on test_rest_api.py. It take a little bit longer as the unit test migration force me to get deeper into the understanding NEP5 standard and transfering token between different accounts. I have already deploy similar contract on the privnet and trying to migrate additional tests. 

@jseagrave21 , could you please give an update from your side? Thanks.

**Comment by jseagrave21** [comment]
@dauTT I am having similar issues. I have been struggling with understanding what the tests are trying to do and what the error codes mean. If you are having to deploy additional contacts, will you share an updated privnet for me to use? Also maybe some changes in the tests that have resulted in success?

**Comment by dauTT** [comment]
Hi @jseagrave21 ,

I have created a new image: 

```docker pull dautt/neo-privnet-unittest:v0.0.4```

So far I have deployed a couple of SM on the privnet. You can find them on the location 

`root/.neopython/UnitTest-SM` of the running docker container.

I have modified the test module, [test_rest_api.py](https://github.com/CityOfZion/neo-python/tree/master/neo/api/REST/test_rest_api.py) as follow:

1) [test_rest_api2_Dau.txt](https://github.com/CityOfZion/neo-python/files/2279724/test_rest_api2_Dau.txt)
2) [BlockchainFixtureTestCase_Dau.txt](https://github.com/CityOfZion/neo-python/files/2279726/BlockchainFixtureTestCase_Dau.txt)

In order for you to run the test case you will need the following fixtures which are basically create from the above docker image:

[fixtures_v8.tar.gz](https://github.com/CityOfZion/neo-python/files/2279719/fixtures_v8.tar.gz)
[notif_fixtures_v8.tar.gz](https://github.com/CityOfZion/neo-python/files/2279720/notif_fixtures_v8.tar.gz)

Relevant transactions used in the test case have been recorded in the following attached file:

[test relevant transactions.txt](https://github.com/CityOfZion/neo-python/files/2279718/test.relevant.transactions.txt)

I am very close to finish the migration of the [test_rest_api.py](https://github.com/CityOfZion/neo-python/tree/master/neo/api/REST/test_rest_api.py). Hopefully, It will done one or two day.

Let me know if you need further help.

**Comment by jseagrave21** [comment]
@dauTT Awesome! I am digging in!

**Comment by jseagrave21** [comment]
@dauTT I am tying to test my version of test_json_invoke_rpc_api.py but when I do I get this error from the BlockchainFixtureTestCase_Dau
```
ERROR: setUpClass (neo.api.JSONRPC.test_json_invoke_rpc_api.JsonRpcInvokeApiTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/c/users/jseag/neo/neo-python-travis/neo/Utils/BlockchainFixtureTestCase_Dau.py", line 35, in setUpClass
    response.raise_for_status()
  File "/mnt/c/users/jseag/Neo/neo-python-travis/venv/share/python-wheels/requests-2.18.4-py2.py3-none-any.whl/requests/models.py", line 935, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 403 Client Error: Forbidden for url: https://s3.us-east-2.amazonaws.com/cityofzion/fixtures/fixtures_v8.tar.gz```

Am I doing something wrong or do you need to give me access to the new fixture file?

**Comment by dauTT** [comment]
Hi @jseagrave21 ,

if you look how the test cases are setup in test_rest_api2_Dau.py or test_rest_api2.py,

```@classmethod
    def setUpClass(cls):
 
        super(NotificationDBTestCase, cls).setUpClass()
 
        if not os.path.exists(cls.N_FIXTURE_FILENAME):
            logzero.logger.info(
                "downloading fixture notification database from %s. this may take a while" % cls.N_FIXTURE_REMOTE_LOC)
 
            response = requests.get(cls.N_FIXTURE_REMOTE_LOC, stream=True)
 
            response.raise_for_status()
            with open(cls.N_FIXTURE_FILENAME, 'wb+') as handle:
                for block in response.iter_content(1024):
                    handle.write(block)
```

you can avoid to download the fixtures from AWS by putting the fixture archives here : 

```os.path.join(settings.DATA_DIR_PATH, 'docker-privnet/Chains/') ```

From my understanding, after we have migrated all the test cases to the privnet, we will create new  fixture archives and ask @localhuman to upload them to AWS.  Until then we have to do a workaround.

**Comment by jseagrave21** [comment]
@dauTT Okay, I am almost done with test_json_invoke_rpc_api.py but I am running into an issue with test_invoke_5. I think it is because I am trying to use the wrong script for the invokescript. Where does the invokescript come from? Is it something I need to generate? Specifically, I am referring to [here](https://github.com/CityOfZion/neo-python/blob/ba815b870614d01f9295d934c8dc3a172765e1aa/neo/api/JSONRPC/test_json_invoke_rpc_api.py#L139)

**Comment by dauTT** [comment]
Thanks @jseagrave21 , I will review your test cases soon and give you feedback.

@ixje , I need your help here to understand this strange behavior in test 9 of this module, [test_rest_api.py](https://github.com/CityOfZion/neo-python/tree/master/neo/api/REST/test_rest_api.py)

``` 
def test_9_by_tx(self):
        mock_req = requestMock(path=b'/tx/0x4c927a7f365cb842ea3576eae474a89183c9e43970a8509b23570a86cb4f5121')
        res = self.app.get_by_tx(mock_req, '0x4c927a7f365cb842ea3576eae474a89183c9e43970a8509b23570a86cb4f5121')
        jsn = json.loads(res)
        self.assertEqual(jsn['total'], 1)
        results = jsn['results']
        self.assertEqual(len(results), 1)
```
the transaction ` tx = 4c927a7f365cb842ea3576eae474a89183c9e43970a8509b23570a86cb4f5121 ` is included in `block 455553`. You can see more information about this test case here:
1) http://testnotifications.neeeo.org/v1/transaction/4c927a7f365cb842ea3576eae474a89183c9e43970a8509b23570a86cb4f5121

2) http://testnotifications.neeeo.org/v1/notifications/block/455553

If you look at the `get_by_tx` method you will see that the notification block (`block_notifications`) is actually stored at  

``` height -1 = 455553-1 = 455552 ```  

```
def get_by_tx(self, request, tx_hash):
        request.setHeader('Content-Type', 'application/json')

        bc = Blockchain.Default()  # type: Blockchain
        notifications = []
        try:
            hash = UInt256.ParseString(tx_hash)
            tx, height = bc.GetTransaction(hash)
            print (tx, ' height =', height)
            if not tx:
                return self.format_message("Could not find transaction for hash %s" % (tx_hash))
            block_notifications = self.notif.get_by_block(height - 1)
            print ("block_notifications:", block_notifications)
            for n in block_notifications:
                if n.tx_hash == tx.Hash:
                    notifications.append(n)
        except Exception as e:
            logger.info("Could not get tx with hash %s because %s " % (tx_hash, e))
            return self.format_message("Could not get tx with hash %s because %s " % (tx_hash, e))

        return self.format_notifications(request, notifications)
```
**In particular we can notice that the test transaction is stored at block 455553 but the corresponding notifications block is at 455552**

If I use my fixtures and consider the following transaction which happened at block 9583 :
```
neo> wallet tkn_send b9fbcff6e50fd381160b822207231233dd3c56c2 AK2nJJpJr6o664CWJK
i1QRXjqeic2zRp8y AXpNr3SDfLXbPHNdqxYeHK5cYpKMHZxMZ9 10000                       
Used 2.807 Gas 

-----------------------------------------------------------
Will transfer 10000.00000000 NXT2 from AK2nJJpJr6o664CWJKi1QRXjqeic2zRp8y to AXpNr3SDfLXbPHNdqxYeHK5cYpKMHZxMZ9
Transfer fee: 0.0001 
-------------------------------------------------------------

[Password]> ***                                                                 
[I 180809 23:50:52 Transaction:615] Verifying transaction: b'a2a37fd2ab7048d70d51eaa8af2815e0e542400329b05a34274771174180a7e8' 
Relayed Tx: a2a37fd2ab7048d70d51eaa8af2815e0e542400329b05a34274771174180a7e8 
```
and use it in test 9: 
```
def test_9_by_tx(self):
        mock_req = requestMock(path=b'/tx/0xa2a37fd2ab7048d70d51eaa8af2815e0e542400329b05a34274771174180a7e8')
        res = self.app.get_by_tx(mock_req, '0xa2a37fd2ab7048d70d51eaa8af2815e0e542400329b05a34274771174180a7e8')
        jsn = json.loads(res)
        self.assertEqual(jsn['total'], 1)
        results = jsn['results']
        self.assertEqual(len(results), 1)

```
**the test fail because the `block_notifactions` ( related to `Tx: a2a37fd2ab7048d70d51eaa8af2815e0e542400329b05a34274771174180a7e8 `) are stored at 9583 instead of 9582 = 9583 -1. So in this particular both the the block of the test transaction and the corresponding notifications block have the same height number, 9583 **

```
{'type': 'SmartContract.Runtime.Notify', 
'contract': '0xb9fbcff6e50fd381160b822207231233dd3c56c2', 
'block': 9583, 
'tx': '0xa2a37fd2ab7048d70d51eaa8af2815e0e542400329b05a34274771174180a7e8', 
'notify_type': 'transfer', 
'addr_to': 'AXpNr3SDfLXbPHNdqxYeHK5cYpKMHZxMZ9', 
'addr_from': 'AK2nJJpJr6o664CWJKi1QRXjqeic2zRp8y', 'amount': 1000000000000}

```

Do you have any explanation for this strange behavior? 

(more information about the the `tx = a2a37fd2ab7048d70d51eaa8af2815e0e542400329b05a34274771174180a7e8` can be found in the fixture archives or in the docker image: `docker pull dautt/neo-privnet-unittest:v0.0.4`)

**Comment by ixje** [comment]
Very good question and no idea. I've asked the author & maintainer if they can explain why that `-1` block index is there in the first place. Once I understand that I can actually think about or look into why it's different now.

**Comment by dauTT** [comment]
Thanks @ixje  for solving this puzzle. :+1: 

I have cross-check again using the latest notification file:

```"NotificationBootstrapFile":"https://s3.us-east-2.amazonaws.com/cityofzion/bootstrap_testnet/TestNotif1413xxx.tar.gz"```

Indeed the  **notification block** of the transaction, ```0x4c927a7f365cb842ea3576eae474a89183c9e43970a8509b23570a86cb4f5121 ```, 
is **455553** as we would have expected. So nothing to worry and we can move on with the migration of the unit tests.

@jseagrave21, I will try review everything by the end of this week.

**Comment by dauTT** [comment]
@jseagrave21 , I have  reviewed the following:

- test_json_invoke_rpc_api.py (Good. We still need to fix `test_invoke_5`)
- test_json_rpc_api.py (Still some work to do)
```
----------------------------------------------------------------------
Ran 48 tests in 18.827s

FAILED (failures=15, errors=1)
```
In regard to your question related to `test_get_block_int`, I still need to investigate further.

- test_smart_contract3.py (Good. :+1: )
- test_smart_contract2.py (Good :+1: )

**Comment by dauTT** [comment]
@ixje ,

In order for me to migrate this [test case](https://github.com/CityOfZion/neo-python/blob/master/neo/api/REST/test_rest_api.py#L180),
```
def test_pagination_for_addr_results(self):
        mock_req = requestMock(path=b'/addr/AFmseVrdL9f9oyCzZefL9tG6UbvhPbdYzM')
        res = self.app.get_by_addr(mock_req, 'AFmseVrdL9f9oyCzZefL9tG6UbvhPbdYzM')
        jsn = json.loads(res)
        self.assertEqual(jsn['total'], 1027)
        results = jsn['results']
        self.assertEqual(len(results), 500)
        self.assertEqual(jsn['total_pages'], 3)

        mock_req = requestMock(path=b'/addr/AFmseVrdL9f9oyCzZefL9tG6UbvhPbdYzM?page=1')
        res = self.app.get_by_addr(mock_req, 'AFmseVrdL9f9oyCzZefL9tG6UbvhPbdYzM')
        jsn = json.loads(res)
        self.assertEqual(jsn['total'], 1027)
        results = jsn['results']
        self.assertEqual(len(results), 500)

        mock_req = requestMock(path=b'/addr/AFmseVrdL9f9oyCzZefL9tG6UbvhPbdYzM?page=2')
        res = self.app.get_by_addr(mock_req, 'AFmseVrdL9f9oyCzZefL9tG6UbvhPbdYzM')
        jsn = json.loads(res)
        self.assertEqual(jsn['total'], 1027)
        results = jsn['results']
        self.assertEqual(len(results), 500)

        mock_req = requestMock(path=b'/addr/AFmseVrdL9f9oyCzZefL9tG6UbvhPbdYzM?page=3')
        res = self.app.get_by_addr(mock_req, 'AFmseVrdL9f9oyCzZefL9tG6UbvhPbdYzM')
        jsn = json.loads(res)
        self.assertEqual(jsn['total'], 1027)
        results = jsn['results']
        self.assertEqual(len(results), 27)
```
 I will need to to create more then 1000 transfers. I am doing that automatically taking for granted that [issue #558](  https://github.com/CityOfZion/neo-python/issues/558) will be fixed. In case issue # 558 is not a bug but a feature, I won't be able to generate the test case automatically. In that case we will need to do a lot of manual work if we want to preserve the test case. :(

**Comment by jseagrave21** [comment]
@dauTT Thank you for the review. When I asked @localhuman about the `invokeScript` he sent me [here](https://github.com/CityOfZion/neo-python/blob/master/neo/Wallets/NEP5Token.py#L92) but I am not sure how to implement this to build the script in order to retrieve the information to satisfy `test_invoke_5` as written. Could you take a look and maybe we can discuss on Discord (in real time). It should be simple but I don't understand enough to make it work.

**Comment by dauTT** [comment]
@jseagrave21 , I will have a look at `test_invoke_5` in the upcoming days. If needed we can arrange a video call.

From my side, I manage to fully migrate [test_rest_api.py](https://github.com/CityOfZion/neo-python/tree/master/neo/api/REST/test_rest_api.py) to the privnet. 

Please use the new update fixtures archives and image:

[fixtures_v8.tar.gz](https://github.com/CityOfZion/neo-python/files/2304114/fixtures_v8.tar.gz)
[notif_fixtures_v8.tar.gz](https://github.com/CityOfZion/neo-python/files/2304117/notif_fixtures_v8.tar.gz)
```docker pull dautt/neo-privnet-unittest:v0.0.18```

Relevant transactions added in the new image are to be found in the following attachment:

[test relevant transactions 3.txt](https://github.com/CityOfZion/neo-python/files/2304129/test.relevant.transactions.3.txt)

@ixje, we are getting close to the completion of this task. 

**To summarize the next steps:**
1) Dau will look at `test_invoke_5`
2) @jseagrave21 will complete the migration of `test_json_rpc_api.py`
3) Dau will integrate @jseagrave21  work and make sure that all test cases are running correctly
4) @ixje  will do the final review
5) ask @localhuman  to upload the new fixture archives to AWS.
6) documentation 
7) close the task

**Comment by jseagrave21** [comment]
@dauTT I like your plan. My only question at this point is the issue I saw with `test_get_block_int`. When I query block 10 I return block 9 as indicated in my previous comment. Have you investigated this? Do I need to create a separate issue from the REST API -1 issue?

**Comment by jseagrave21** [comment]
I did some more testing and I think the right script for `test_invoke_5` is `0d444601ee23a8882f6fce312d5a0b59ea02f323f7f391e6ef3341669f411cbe`. However, this does not return the expected values (i.e. bytearray(b'NEX Template V2'), bytearray(b'NXT2'), 8). This its raw return before it is sent to `ContractParameter.FromJson`:
```
'stack': [{'type': 'ByteArray', 'value': '444601ee23a8882f6fce312d5a'}, {'type': 'ByteArray', 'value': '59ea02f323f7f391e6ef33'}, {'type': 'ByteArray', 'value': '669f411cbe'}]
```
Hopefully this helps.

**Comment by dauTT** [comment]
Thanks @jseagrave21  for the hint! :+1: 

`test_invoke_5` fixed:

```
def test_invoke_5(self):
        test_script = "00046e616d6567c2563cdd3312230722820b1681d30fe5f6cffbb9000673796d626f6c67c2563cdd3312230722820b1681d30fe5f6cffbb90008646563696d616c7367c2563cdd3312230722820b1681d30fe5f6cffbb9" 
        
        req = self._gen_rpc_req("invokescript", params=[test_script])
        mock_req = mock_request(json.dumps(req).encode("utf-8"))
        res = json.loads(self.app.home(mock_req))
        self.assertEqual(res['result']['state'], VMStateStr(VMState.HALT + VMState.BREAK))

        results = []
        for p in res['result']['stack']:
            results.append(ContractParameter.FromJson(p))
            
        self.assertEqual(len(results), 3)
        self.assertEqual(results[0].Value, bytearray(b'NEX Template V2'))
        self.assertEqual(results[1].Value, bytearray(b'NXT2'))
        self.assertEqual(results[2].Value, 8)
```

**Comment by jseagrave21** [comment]
Wow. I spent a loong time trying to figure out how to do that! Good work @dauTT ! I am continuing to work on my remaining test but I have a feeling I am not going to be able to complete all of it. I will get as far as I can and then turn it over to you to look at.

**Comment by dauTT** [comment]
@jseagrave21 ,

I looked at the test case [`test_get_block_int`](https://github.com/jseagrave21/neo-python/blob/09877fcf4a488154b59b819ab13d65bd4c4fce70/neo/api/JSONRPC/test_json_rpc_api.py#L172) in your branch:

```
def test_get_block_int(self):
        req = self._gen_rpc_req("getblock", params=[10, 1])
        mock_req = mock_request(json.dumps(req).encode("utf-8"))
        res = json.loads(self.app.home(mock_req))
        self.assertEqual(res['result']['index'], 9)
        self.assertEqual(res['result']['hash'], '0x7c5b4c8a70336bf68e8679be7c9a2a15f85c0f6d0e14389019dcc3edfab2bb4b')
        self.assertEqual(res['result']['confirmations'], 9598)
        self.assertEqual(res['result']['nextblockhash'], '0x7c5b4c8a70336bf68e8679be7c9a2a15f85c0f6d0e14389019dcc3edfab2bb4b')
```

**This test case is not correct because we expect index = 10 instead of 9**. However the test case pass why? the reason why it pass has nothing to do with #554. In general, at the present moment, when running our test cases by default it takes the setting of the `TEST` environment. So, the all privnet blockchain get the extra genesis block of `TEST`and therefore the chain is off of one. This is not correct. **To fix this issue we need to activate the setting of privnet before we run the test cases**. So please use this new update file going forward:

[BlockchainFixtureTestCase_Dau.py.txt](https://github.com/CityOfZion/neo-python/files/2312092/BlockchainFixtureTestCase_Dau.py.txt)

Fortunately this change should not impact our previous testing. So, the new update `test_get_block_int`
will look like this:

```
def test_get_block_int(self):
        req = self._gen_rpc_req("getblock", params=[10, 1])
        mock_req = mock_request(json.dumps(req).encode("utf-8"))
        res = json.loads(self.app.home(mock_req))
        print(res)
        self.assertEqual(res['result']['index'], 10)
        self.assertEqual(res['result']['hash'], '0xd69e7a1f62225a35fed91ca578f33447d93fa0fd2b2f662b957e19c38c1dab1e')
        self.assertEqual(res['result']['confirmations'], 12201)
        self.assertEqual(res['result']['nextblockhash'], '0x2b1c78633dae7ab81f64362e0828153079a17b018d779d0406491f84c27b086f')
```

I notice that the following assetion is no useful for us anymore:

```self.assertEqual(res['result']['confirmations'], 12201)```

If we look at the method, [get_tx_output](https://github.com/CityOfZion/neo-python/blob/master/neo/api/JSONRPC/JsonRpcApi.py#L284), behind the `confirmations` attribute we see that it is calculated as follow:

```
jsn['confirmations'] = Blockchain.Default().Height - header.Index + 1
```

This is not a good check in my opinion because every time we add a new block to the privnet fixtures, this confirmations number will change and therefore the assertion will fail. @ixje , are you fine to remove the above assertion from `test_get_block_int`?

**Comment by jseagrave21** [comment]
@dauTT Perfect! That answers my question exactly. I am glad they were unrelated issues! I am still working but I haven't had much time this week because of school. Let me know if you decide to handle the remaining portion of `test_json_rpc_api.py` yourself.

**Comment by dauTT** [comment]
@jseagrave21  no worry. I understand as I have also limited capacity due to work. So keep your pace. 👍

**Comment by jseagrave21** [comment]
@dauTT For ` test_account_state_not_existing_yet`, I need an empty wallet. Is there an empty wallet on your testcase?

**Comment by dauTT** [comment]
@jseagrave21 ,

the existing [test case](https://github.com/CityOfZion/neo-python/blob/master/neo/api/JSONRPC/test_json_rpc_api.py#L117) seems to work because this account, `AHozf8x8GmyLnNv8ikQcPKgRHQTbFi46u2` does not exist on the privnet image (`dautt/neo-privnet-unittest:v0.0.18`)
```
def test_account_state_not_existing_yet(self):
        addr_str = 'AHozf8x8GmyLnNv8ikQcPKgRHQTbFi46u2'
        req = self._gen_rpc_req("getaccountstate", params=[addr_str])
        mock_req = mock_request(json.dumps(req).encode("utf-8"))
        res = json.loads(self.app.home(mock_req))
        self.assertEqual(res['result']['balances'], {})
        self.assertEqual(res['result']['script_hash'], addr_str)
```
So we don't need to create an empty wallet.

**Comment by jseagrave21** [comment]
@dauTT okay, I didn't know if that accomplished the intent of the test. Thank you for looking at it!

**Comment by ixje** [comment]
I see so much effort and perseverance, keep it up! 💪 
If there's anything I can do or clarify let me know. I might be a slow in responding as my online time is a bit on and off the last week and probably next week.

**Comment by jseagrave21** [comment]
@dauTT I think I have migrated as much as I can figure out. Here is my latest test:
https://github.com/jseagrave21/neo-python/blob/patch-5/neo/api/JSONRPC/test_json_rpc_api.py

Here are the failures/error I am still getting:
```
======================================================================
ERROR: test_send_raw_tx (neo.api.JSONRPC.test_json_rpc_api.JsonRpcApiTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/c/users/jseag/neo/neo-python-travis/neo/api/JSONRPC/test_json_rpc_api.py", line 425, in test_send_raw_tx
    self.assertEqual(res['result'], False)
KeyError: 'result'

======================================================================
FAIL: test_get_storage_item (neo.api.JSONRPC.test_json_rpc_api.JsonRpcApiTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/c/users/jseag/neo/neo-python-travis/neo/api/JSONRPC/test_json_rpc_api.py", line 342, in test_get_storage_item
    self.assertEqual(res['result'], '00f938dbba05')
AssertionError: None != '00f938dbba05'

======================================================================
FAIL: test_get_storage_item2 (neo.api.JSONRPC.test_json_rpc_api.JsonRpcApiTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/c/users/jseag/neo/neo-python-travis/neo/api/JSONRPC/test_json_rpc_api.py", line 352, in test_get_storage_item2
    self.assertEqual(res['result'], '003a0d7914b200')
AssertionError: None != '003a0d7914b200'

======================================================================
FAIL: test_get_storage_item_bad_contract_hash (neo.api.JSONRPC.test_json_rpc_api.JsonRpcApiTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/c/users/jseag/neo/neo-python-travis/neo/api/JSONRPC/test_json_rpc_api.py", line 376, in test_get_storage_item_bad_contract_hash
    self.assertTrue('error' in res)
AssertionError: False is not true

----------------------------------------------------------------------
Ran 48 tests in 9.474s

FAILED (failures=3, errors=1)
```

I am not sure if your contract has storage and if it does, how to pull out the required info. And, I don't know how to send a valid raw tx. cc @ixje

**Comment by dauTT** [comment]
Thanks @jseagrave21  for your hard work!  :+1: 
I will have a look at the remaining test cases tomorrow.

**Comment by dauTT** [comment]
Hi @ixje ,

there are few test cases left. This [one]( https://github.com/CityOfZion/neo-python/blob/master/neo/api/JSONRPC/test_json_rpc_api.py#L337) is really tough:

```
    def test_get_storage_item(self):
        contract_hash = '16f1559c3c27d66d087bef936804105457617c8a'
        storage_key = binascii.hexlify(b'totalSupply').decode('utf-8')
        req = self._gen_rpc_req("getstorage", params=[contract_hash, storage_key])
        mock_req = mock_request(json.dumps(req).encode("utf-8"))
        res = json.loads(self.app.home(mock_req))
        self.assertEqual(res['result'], '00f938dbba05')
        actual_val = int.from_bytes(binascii.unhexlify(res['result'].encode('utf-8')), 'little')
        self.assertEqual(actual_val, 6300100000000)
```

I have try to replicate the test case on the privnet by running this transaction:

```
neo> testinvoke b9fbcff6e50fd381160b822207231233dd3c56c2 totalSupply []         
Used 0.32 Gas 

-------------------------------------------------------------------------------------------------------------------------------------
Test invoke successful
Total operations: 347
Results [{'type': 'ByteArray', 'value': '00a031a95fe300'}]
Invoke TX GAS cost: 0.0
Invoke TX fee: 0.0001
-------------------------------------------------------------------------------------------------------------------------------------

Enter your password to continue and invoke on the network

[password]> **********                                                          
[I 180830 23:01:20 Transaction:618] Verifying transaction: b'bad4face9c794900569f861a0386ccea1c3d99fda5c3602ab9d213d981524fe2' 
Relayed Tx: bad4face9c794900569f861a0386ccea1c3d99fda5c3602ab9d213d981524fe2 
```
However the transaction does not produce the desire 
```
storage_key = binascii.hexlify(b'totalSupply').decode('utf-8')
```

If I look at all the key in my privnet:
```
def test_storage():
    from neo.Implementations.Blockchains.LevelDB.DBCollection import DBCollection
    from neo.Implementations.Blockchains.LevelDB.DBPrefix import DBPrefix
    from neo.Core.State.StorageItem import StorageItem
    import binascii 
    blockchain = LevelDBBlockchain("/home/dau/.neopython/docker-privnet/Chains/backup/fixtures/test_chain") 
    
    Blockchain.RegisterBlockchain(blockchain)
    _db = Blockchain.Default()._db
    sn=_db.snapshot()
    storages = DBCollection(_db, sn, DBPrefix.ST_Storage, StorageItem)
    for k in storages.Keys:
        value =storages.TryGet(k).Value
        print(k, "\n", value, "\n new k, v \n")
```
Output:
```
new k, v 

 b'\xc2V<\xdd3\x12#\x07"\x82\x0b\x16\x81\xd3\x0f\xe5\xf6\xcf\xfb\xb9H\x03;X\xefT|\xbfT\xc8\xee/r\xa4-[`<\x00\xaf' 
 b'\x00B\xe4\xb52' 

 new k, v 

 b'\xc2V<\xdd3\x12#\x07"\x82\x0b\x16\x81\xd3\x0f\xe5\xf6\xcf\xfb\xb9in_circulation' 
 b'\x00\xa01\xa9_\xe3\x00' 

 new k, v 

 b'\xc2V<\xdd3\x12#\x07"\x82\x0b\x16\x81\xd3\x0f\xe5\xf6\xcf\xfb\xb9initialized' 
 b'\x01' 
```
they are ending either  with no "human readable" suffix or with suffix  in ` ['in_circulation', 'initialized']`. Do you have any hint how I can generate a transaction with storage key ending with suffix `totalSupply`?.

I believe the answer of this question may have to do with the [ico_template](https://github.com/CityOfZion/neo-boa/blob/master/boa_test/example/demo/ICO_Template.py#L20):
```
NEP5_METHODS = ['name', 'symbol', 'decimals', 'totalSupply', 'balanceOf', 'transfer', 'transferFrom', 'approve', 'allowance']
```

**Comment by jseagrave21** [comment]
@dauTT Nice to see it coming together!

**Comment by ixje** [comment]
@dauTT super nice! I'm excited to go through all this, ~14 MB sounds amazing 🥇 
As for the 2 tests, yes let's keep those in. It's a small adjustment to make whenever new tests are added, and still useful to quickly verify proper loading of the right blockchain fixtures. I might be able to go through them on Wednesday, otherwise no sooner then Friday (just to give an indication)

**Comment by ixje** [comment]
**test cases**
@dauTT the tests in your link look good to me , great job (@jseagrave21 too)👌 
You probably want to run `make lint` before submitting the PR because I think i see some spaces that violate the automatically enforced code style checker. 

**fixture testnet data**
The collection of smart contracts in the zip also looks fine. The readme can use a little bit of work. I know what you did and i can make sense of most data in the readme, but i think a first time reader might not or have trouble following. I think a structure along the following lines with a bit more words would be great
```
Intro
- neo-python fixtures are based of a private unit test net <link>
- it currently contains the following smart contracts ...
- a default wallet with the following credentials is in place...

How to add new fixture tests
- pull & run private net (howto...)
- use wallet with the following credentials as needed
- add test, pr, ask for image update
 ```

For instructions like
```
rm -rf /home/dau/.neopython/Chains/privnet
```
try to make them generic / applicable to all readers i.e.
 ```
rm -rf ~/.neopython/Chains/privnet
 ```

**file placing for PR**
As the `BlockChainFixtureTestCase` resides under the directory `neo/Utils` I think the right place to put the contents of `UnitTest-SM.zip` could be something like `neo/Utils/fixtures-testnet/`

The fixture files should have been uploaded to CoZ AWS by now so you should be good to PR and have travis pass (if not then i'll restart the build once they're up, no worries). I'll have to discuss how we want to port/backup the docker image but we can at least proceed for now.

**Comment by localhuman** [comment]
S3 Files:
https://s3.us-east-2.amazonaws.com/cityofzion/fixtures/fixtures_v8.tar.gz
https://s3.us-east-2.amazonaws.com/cityofzion/fixtures/notif_fixtures_v8.tar.gz
https://s3.us-east-2.amazonaws.com/cityofzion/fixtures/UnitTest_SM_v8.zip

Thanks for all the hard work on this, I can't imagine yet how much more enjoyable testing will be 🥇

**Comment by jseagrave21** [comment]
@dauTT let me know if you need any help with anything else before submitting the PR. Or if you would like me to review anything. I am glad I could help and would be happy to in any way I can. Great work!

**Comment by dauTT** [comment]
Thanks @jseagrave21 for the help.
I am a little bit confuse about submitting the PR.

@ixje / @jseagrave21, **Shall I submit a PR to the Development branch or to the Master branch?**
(If it is the Development branch, we need perhaps to update the test cases again before submitting a PR)

If I run, one by one, the test modules which we have migrated to the privnet, they are all passing using the latest version of neo-python  `v0.7.8`. However when I run this command:

`python -m unittest discover neo`

Some of them are failing and there are too many failures and errors.
```
Ran 353 tests in 86.778s
FAILED (failures=17, errors=11)
```
It seems there are still come conflicts which need to resolve. @jseagrave21, if you have time you can also look at this and let me know what is your opinion. Thanks!

**Comment by jseagrave21** [comment]
@dauTT We are supposed to submit the PR to the Development Branch.

**Comment by jseagrave21** [comment]
Can you link to or share the report so we can see what the errors are? Maybe a Google Doc?

**Comment by dauTT** [comment]
Hi @jseagrave21 ,
I have this log file: [log.txt](https://github.com/CityOfZion/neo-python/files/2363469/log.txt),

My first impression is that when we run all the tests together, the fixtures are coming not from the privnet but from the testnet. Need to investigate further...

**Comment by jseagrave21** [comment]
@dauTT Here are the results from my tests:

[Log1.txt](https://github.com/CityOfZion/neo-python/files/2363480/Log1.txt)

I noticed in test_json_rpc_api.py that the same issue where `getblock` returns the wrong block has returned.

**Comment by jseagrave21** [comment]
@dauTT When I was building my environment, I created a fresh pull of the CoZ Development branch and then modified all the files you mention [here](https://github.com/CityOfZion/neo-python/issues/478#issuecomment-418194388) by copy and pasting from your neo-python repository "latest...." branch. I did not download anything else because I was assuming that since the files have been uploaded to AWS, then we should be able to pull down from there.

I also didn't do anything with the contents of `UnitTest-SM.zip`. Please let me know if I need to take further action to create an accurate environment.

**Comment by jseagrave21** [comment]
Also, since several of the errors involve `UserWalletTestCase`, should that have been updated?

**Comment by jseagrave21** [comment]
@dauTT I think option 1 is best. But, even if we go ahead with the PR, they will not approve it until we have a passing build.

**Comment by jseagrave21** [comment]
That is why @ixje has delayed some of the other PRs. Because some are dependent on the `WalletFixtureTestCase`.

**Comment by jseagrave21** [comment]
@dauTT I don't think it will be too hard to fix the errors but I am concerned about the errors we have already fixed that I mentioned earlier (`getblock` returning 9 when 10 was queried)

**Comment by dauTT** [comment]
@jseagrave21 ,
in regards to the conflict of fixtures that I was talking above it is related to the fact that some other test cases (e.g, `VerifiableTestCase`) which are not related to the `BlockchainFixtureTestCase` class are using the location 
```
os.path.join(settings.DATA_DIR_PATH, 'fixtures/test_chain')
```
For some strange reason the  `setup` and `teardown` of some test cases didn't work properly and in the end the fixtures are mixed up. The mixing of fixtures causes errors we have already fixed (e.g, `getblock` returning 9 when 10 was queried). This can be resolved by tweaking `BlockchainFixtureTestCase` but it will better to fix the problem at the source.

To my knowledge the following is the updated list of test cases to migrate to the privnet. These test cases depend on `BlockchainFixtureTestCase` and `WalletFixtureTestCase`: 

0. [test_withdraw.py](https://github.com/CityOfZion/neo-python/tree/master/neo/contrib/nex/test_withdraw.py)
1. [test_LevelDBBlockchain.py](https://github.com/CityOfZion/neo-python/tree/master/neo/Implementations/Blockchains/LevelDB/test_LevelDBBlockchain.py) (Done, potentially to update)
2. [test_leveldb.py](https://github.com/CityOfZion/neo-python/tree/master/neo/Implementations/Blockchains/LevelDB/tests/test_leveldb.py) (Done, potentially to update)
3. [test_user_wallet.py](https://github.com/CityOfZion/neo-python/tree/master/neo/Implementations/Wallets/peewee/test_user_wallet.py)
4. [test_create_user_wallet.py](https://github.com/CityOfZion/neo-python/tree/master/neo/Implementations/Wallets/peewee/test_create_user_wallet.py)
5. [test_interop_blockchain.py](https://github.com/CityOfZion/neo-python/tree/master/neo/VM/tests/test_interop_blockchain.py) (Done, potentially to update)
6. [test_rest_api.py](https://github.com/CityOfZion/neo-python/tree/master/neo/api/REST/test_rest_api.py) (Done, potentially to update)
7. [test_json_invoke_rpc_api.py](https://github.com/CityOfZion/neo-python/tree/master/neo/api/JSONRPC/test_json_invoke_rpc_api.py) (Done, potentially to update)
8. [test_json_rpc_api.py](https://github.com/CityOfZion/neo-python/tree/master/neo/api/JSONRPC/test_json_rpc_api.py) (Done, potentially to update)
9. [test_token_commands.py](https://github.com/CityOfZion/neo-python/tree/master/neo/Prompt/Commands/tests/test_token_commands.py)
10. [test_send_command.py](https://github.com/CityOfZion/neo-python/tree/master/neo/Prompt/Commands/tests/test_send_command.py)
11. [test_wallet_commands.py](https://github.com/CityOfZion/neo-python/tree/master/neo/Prompt/Commands/tests/test_wallet_commands.py)
12. [test_claim_command.py](https://github.com/CityOfZion/neo-python/tree/master/neo/Prompt/Commands/tests/test_claim_command.py)
13. [test_node_leader.py](https://github.com/CityOfZion/neo-python/tree/master/neo/Network/test_node_leader.py)
14. [test_smart_contract.py](https://github.com/CityOfZion/neo-python/tree/master/neo/SmartContract/tests/test_smart_contract.py) (Done, potentially to update)
15. [test_payable.py](https://github.com/CityOfZion/neo-python/tree/master/neo/SmartContract/tests/test_payable.py)
16. [test_breakpoints.py](https://github.com/CityOfZion/neo-python/tree/master/neo/SmartContract/tests/test_breakpoints.py)
17. [test_storage_find.py](https://github.com/CityOfZion/neo-python/tree/master/neo/SmartContract/tests/test_storage_find.py)
18. [test_smart_contract3.py](https://github.com/CityOfZion/neo-python/tree/master/neo/SmartContract/tests/test_smart_contract3.py) (Done, potentially to update)
19. [test_smart_contract2.py](https://github.com/CityOfZion/neo-python/tree/master/neo/SmartContract/tests/test_smart_contract2.py) (Done, potentially to update)
20. [test_gas_costs.py](https://github.com/CityOfZion/neo-python/tree/master/neo/SmartContract/tests/test_gas_costs.py)
21. [test_migrate_destroy.py](https://github.com/CityOfZion/neo-python/tree/master/neo/SmartContract/tests/test_migrate_destroy.py)

So there are still some work to do. @jseagrave21 , shall I assign half of the remaining test cases to me and half to you?

**Comment by jseagrave21** [comment]
@dauTT okay that sounds good!

**Comment by jseagrave21** [comment]
Is there anything that we need help with? Like the tweaking of `BlockchainFixtureTestCase` or fixing it at the source? So we can have three people (including @ixje) working on it?

**Comment by dauTT** [comment]
Let us involve @ixje , for difficult task as we will need his help of course. :smile: 
I have assigned the first items to me an the remaining to you. Thanks!

**Comment by ixje** [comment]
I've asked those with merging rights to not merge any PRs that rely on or inherit from `BlockchainFixtureTestCase` until we get this one PR'ed. What else can i do for now? @dauTT are you on discord? If not, joining could make communicating a bit faster/easier at this stage

**Comment by jseagrave21** [comment]
@dauTT okay, I will start at the bottom and work my way up

**Comment by jseagrave21** [comment]
@dauTT So, `test_claim_command.py` completely fails and I think I know why. The `WalletFixtureTestCase` references wallets which do not exist in your fixture. You need to add the wallets you have been using to the `WalletFixtureTestCase`.

I think the easiest way to do this would be to open `np-prompt` and create a wallet labeled `testwallet.db3` with the password `testpassword` and then import the wif for the address `AK2nJJpJr6o664CWJKi1QRXjqeic2zRp8y` and delete the other (original) address. Then you could replace the `testwallet.db3` which already resided in fixtures/. Then create another wallet labeled `testwallet2.db3` with the password `testwallet` and import the wif for the address `AXpNr3SDfLXbPHNdqxYeHK5cYpKMHZxMZ9` and delete the other (orginal) address. Then you could replace the `testwallet2.db3` in fixtures/. I am not sure if you have a third address you were using. If not, you should make a new wallet to replace `testwallet3.db3` and send some assets to it. If you do this, then we shouldn't have to completely change all code for the `class UserWalletTestCase(BlockchainFixtureTestCase)` But we will have to change a lot. @ixje Could you double check what I am recommending to see if it makes sense? Essentially we are rebuilding the `WalletFixtureTestCase`.

**Comment by ixje** [comment]
Yes I think you need to 'rebuild' the wallets in `WalletFixtureTestCase`. These wallets will rely on blockchain data coming from `fixtures_v7.tar.gz` which is unlikely to match the new `v8` blockchain data.

ps: with 'rebuild' i don't mean the command, but making sure the wallet addresses in the wallet actually exists in the blockchain data of the `v8` fixtures. Including those addresses having funds, transactions and whatever is needed for the tests.

**Comment by dauTT** [comment]
Thanks @jseagrave21 , @ixje. I will do that.

**Comment by jseagrave21** [comment]
@dauTT sounds good. I would love to help with this part but I don't have the WIFs for any of the addresses used and I am not sure of all the addresses which are active.

As soon as the `WalletFixtureTestCase` is migrated, please post a readme with the addresses and their associated wallets and passwords and hashes (Look at test_claim_command.py to see what kind of information is referenced when a `UserWalletTestCase` is set up). The first wallet should have a watch_address set so the second wallet addresses. We might need to update these fixtures to me the demands of particular tests but that will at least get us rolling. 

Afterwards, I will start over to make sure the `UserWalletTestCase` in each test references the correct info and also passes.

**Comment by jseagrave21** [comment]
@dauTT I like what you proposed and I think I see what you are getting at. However, I think we can minimize the number of new wallets needed. Here is my counter-proposal.

[testing.setup.proposal2.docx](https://github.com/CityOfZion/neo-python/files/2372893/testing.setup.proposal2.docx)

**Comment by dauTT** [comment]
@ixje / @jseagrave21 ,
you can find the most updated files here: [link](https://github.com/dauTT/neo-python/tree/latest2-move-unittest-to-privnet-478)

I have modifed the following files: 
```
        modified:   fixtures/testwallet.db3 (autogenerated, no need for review)
	modified:   neo/Implementations/Blockchains/LevelDB/test_LevelDBBlockchain.py (okay)
	modified:   neo/Implementations/Blockchains/LevelDB/tests/test_leveldb.py (okay)
	modified:   neo/Implementations/Wallets/peewee/test_create_user_wallet.py 
	modified:   neo/Implementations/Wallets/peewee/test_user_wallet.py
	modified:   neo/Network/test_node_leader.py 
	modified:   neo/Prompt/Commands/tests/test_claim_command.py (not okay, to investigate why in the fixture we have negative unavailable_bonus!)
	modified:   neo/Prompt/Commands/tests/test_send_command.py
	modified:   neo/Prompt/Commands/tests/test_token_commands.py
	modified:   neo/Prompt/Commands/tests/test_wallet_commands.py
	modified:   neo/Settings.py (okay)
	modified:   neo/SmartContract/tests/StorageTest.avm (autogenerated, no need for review)
	modified:   neo/SmartContract/tests/test_breakpoints.py
	modified:   neo/SmartContract/tests/test_gas_costs.py
	modified:   neo/SmartContract/tests/test_migrate_destroy.py
	modified:   neo/SmartContract/tests/test_smart_contract.py  (okay)
	modified:   neo/SmartContract/tests/test_smart_contract2.py  (okay)
	modified:   neo/SmartContract/tests/test_smart_contract3.py  (okay)
	modified:   neo/Utils/BlockchainFixtureTestCase.py (okay)
	modified:   neo/Utils/VerifiableTestCase.py
	modified:   neo/Utils/WalletFixtureTestCase.py
	modified:   neo/VM/tests/test_interop_blockchain.py (okay)
	modified:   neo/api/JSONRPC/test_json_invoke_rpc_api.py (okay)
	modified:   neo/api/JSONRPC/test_json_rpc_api.py (@ixje, @jsagrave to enrich the test cases as new ones were added in the meantime: test_getbalance_invalid_params, test_validate_address, test_getbalance_invalid_params, test_valid_multirequest, test_multirequest_with_1_invalid_request)
	modified:   neo/api/REST/RestApi.py (okay)
	modified:   neo/api/REST/test_rest_api.py (okay)
	deleted:    neo/contrib/nex/test_withdraw.py (okay, @ixje confirmed)
```
and added the following wallets:
```
fixtures/neo-test-coz.wallet (pwd=coz)
fixtures/neo.test1-bc.wallet (pwd=1234567890)
fixtures/neo-test2-bc.wallet (pwd=1234567890)
fixtures/neo-test1-w.wallet (pwd=1234567890)
fixtures/neo-test2-w.wallet (pwd=1234567890)
fixtures/neo-test3-w.wallet (pwd=1234567890)
```
Latest image created: docker pull dautt/neo-privnet-unittest:v0.0.22

Fixtures to upload to AWS:
[fixtures_v8.tar.gz](https://github.com/CityOfZion/neo-python/files/2386024/fixtures_v8.tar.gz)
[notif_fixtures_v8.tar.gz](https://github.com/CityOfZion/neo-python/files/2386025/notif_fixtures_v8.tar.gz)

(both fixture files are in total about 7MB)
All the test cases have been tested using neo-python `v0.7.9-dev` from the development branch. **They all pass if we execute one test at the time by running these commands**:
```
python -m unittest neo/Implementations/Blockchains/LevelDB/test_LevelDBBlockchain.py
python -m unittest neo/Implementations/Blockchains/LevelDB/tests/test_leveldb.py
python -m unittest neo/Implementations/Wallets/peewee/test_user_wallet.py
python -m unittest neo/Implementations/Wallets/peewee/test_create_user_wallet.py
python -m unittest neo/VM/tests/test_interop_blockchain.py
python -m unittest neo/api/REST/test_rest_api.py
python -m unittest neo/api/JSONRPC/test_json_invoke_rpc_api.py
python -m unittest neo/api/JSONRPC/test_json_rpc_api.py
python -m unittest neo/Prompt/Commands/tests/test_send_command.py
python -m unittest neo/Prompt/Commands/tests/test_wallet_commands.py
python -m unittest neo/Prompt/Commands/tests/test_claim_command.py
python -m unittest neo/Network/test_node_leader.py
python -m unittest neo/SmartContract/tests/test_smart_contract.py
python -m unittest neo/SmartContract/tests/test_payable.py
python -m unittest neo/SmartContract/tests/test_breakpoints.py
python -m unittest neo/SmartContract/tests/test_storage_find.py
python -m unittest neo/SmartContract/tests/test_smart_contract3.py
python -m unittest neo/SmartContract/tests/test_smart_contract2.py
python -m unittest neo/SmartContract/tests/test_gas_costs.py
python -m unittest neo/SmartContract/tests/test_migrate_destroy.py
```
**Hovever if we run this command:**
```
python -m unittest discover neo
```
**we get the following unexpected result:**

```
======================================================================
ERROR: test_initial_setup (api.JSONRPC.test_json_rpc_api.JsonRpcApiTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dau/workspace/python/github.com/neo/dautt/neo-python/neo/api/JSONRPC/test_json_rpc_api.py", line 62, in test_initial_setup
    self.assertTrue(GetBlockchain().GetBlock(0).Hash.To0xString(), '0x996e37358dc369912041f966f8c5d8d3a8255ba5dcbd3447f8a82b55db869099')
AttributeError: 'NoneType' object has no attribute 'Hash'

======================================================================
FAIL: test_add_header (Implementations.Blockchains.LevelDB.tests.test_initial_db.LevelDBTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dau/workspace/python/github.com/neo/dautt/neo-python/neo/Implementations/Blockchains/LevelDB/tests/test_initial_db.py", line 54, in test_add_header
    self.assertEqual(self._blockchain.CurrentHeaderHash, b'd42561e3d30e15be6400b6df2f328e02d2bf6354c41dce433bc57687c82144bf')
AssertionError: b'b3181718ef6167105b70920e4a8fbbd0a0a56aacf460d70e10ba6fa1668f1fef' != b'd42561e3d30e15be6400b6df2f328e02d2bf6354c41dce433bc57687c82144bf'

======================================================================
FAIL: test_block_non_verbose (api.JSONRPC.test_json_rpc_api.JsonRpcApiTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dau/workspace/python/github.com/neo/dautt/neo-python/neo/api/JSONRPC/test_json_rpc_api.py", line 239, in test_block_non_verbose
    self.assertEqual(block.Index, 2003)
AssertionError: 2002 != 2003

======================================================================
FAIL: test_get_block_int (api.JSONRPC.test_json_rpc_api.JsonRpcApiTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dau/workspace/python/github.com/neo/dautt/neo-python/neo/api/JSONRPC/test_json_rpc_api.py", line 182, in test_get_block_int
    self.assertEqual(GetBlockchain().GetBlock(1).ToJson()['hash'], '0x55f745c9098d5d5bdaff9f8f32aad29c904c83d9832b48c16e677d30c7da4273')
AssertionError: '0x996e37358dc369912041f966f8c5d8d3a8255ba5dcbd3447f8a82b55db869099' != '0x55f745c9098d5d5bdaff9f8f32aad29c904c83d9832b48c16e677d30c7da4273'
- 0x996e37358dc369912041f966f8c5d8d3a8255ba5dcbd3447f8a82b55db869099
+ 0x55f745c9098d5d5bdaff9f8f32aad29c904c83d9832b48c16e677d30c7da4273


======================================================================
FAIL: test_get_block_sysfee (api.JSONRPC.test_json_rpc_api.JsonRpcApiTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dau/workspace/python/github.com/neo/dautt/neo-python/neo/api/JSONRPC/test_json_rpc_api.py", line 214, in test_get_block_sysfee
    self.assertEqual(res['result'], 1560)
AssertionError: 1070 != 1560

======================================================================
FAIL: test_getblockhash (api.JSONRPC.test_json_rpc_api.JsonRpcApiTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dau/workspace/python/github.com/neo/dautt/neo-python/neo/api/JSONRPC/test_json_rpc_api.py", line 104, in test_getblockhash
    self.assertEqual(expected_blockhash, res["result"])
AssertionError: '0x049db9f55ac45201c128d1a40d0ef9d4bdc58db97d47d985ce8d66511a1ef9eb' != '0x55f745c9098d5d5bdaff9f8f32aad29c904c83d9832b48c16e677d30c7da4273'
- 0x049db9f55ac45201c128d1a40d0ef9d4bdc58db97d47d985ce8d66511a1ef9eb
+ 0x55f745c9098d5d5bdaff9f8f32aad29c904c83d9832b48c16e677d30c7da4273


======================================================================
FAIL: test_getrawtx_1 (api.JSONRPC.test_json_rpc_api.JsonRpcApiTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dau/workspace/python/github.com/neo/dautt/neo-python/neo/api/JSONRPC/test_json_rpc_api.py", line 322, in test_getrawtx_1
    self.assertEqual(res['blockhash'], '0x6088bf9d3b55c67184f60b00d2e380228f713b4028b24c1719796dcd2006e417')
AssertionError: '0x0e9c0a7e88aeaf217654107041986f4491b447590cfc3d060304cef1b8536ea0' != '0x6088bf9d3b55c67184f60b00d2e380228f713b4028b24c1719796dcd2006e417'
- 0x0e9c0a7e88aeaf217654107041986f4491b447590cfc3d060304cef1b8536ea0
+ 0x6088bf9d3b55c67184f60b00d2e380228f713b4028b24c1719796dcd2006e417


======================================================================
FAIL: test_prompt_open_wallet (bin.test_prompt.PromptTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dau/workspace/python/github.com/neo/dautt/neo-python/neo/bin/test_prompt.py", line 26, in test_prompt_open_wallet
    self.assertIn(checktext, text)
AssertionError: 'Opened' not found in "open wallet fixtures/testwallet.db3\r\ntestpassword\r\n\x1b[32m[I 180915 21:17:39 LevelDBBlockchain:114]\x1b[39m Created Blockchain DB at /home/dau/.neopython/Chains/SC234 \r\n\x1b[32m[I 180915 21:17:46 NotificationDB:73]\x1b[39m Created Notification DB At /home/dau/.neopython/Chains/Test_Notif \r\n\x1b[0m\x1b[?7h\x1b[0;38;5;4mNEO\x1b[0;38;5;10m cli. Type \x1b[0;38;5;197m'help' \x1b[0;38;5;10mto get started\x1b[0m\r\r\n\x1b[0m\r\n\r\n\x1b[?1l\x1b[6n\x1b[?2004h\x1b[?25l\x1b[0m\x1b[?7l\x1b[0m\x1b[Jneo>  \x1b[0m\r\r\n\r\r\n\r\r\n\r\r\n\r\r\n\r\r\n\r\r\n                                                                               \r\x1b[79C \r\x1b[7A\x1b[5C\x1b[?7h\x1b[0m\x1b[?12l\x1b[?25h\x1b[?25l\x1b[?7l\x1b[5Dneo> open wallet fixtures/testwallet.db3                                       \r\x1b[79C \r\x1b[40C\x1b[?7h\x1b[0m\x1b[?12l\x1b[?25h\x1b[?25l\x1b[?7l\x1b[40D\x1b[0m\x1b[Jneo> open wallet fixtures/testwallet.db3                                       \r\x1b[79C \r\x1b[0m\r\r\n\x1b[J\x1b[?7h\x1b[0m\x1b[?12l\x1b[?25h\x1b[?2004l\x1b[?1l\x1b[?2004h\x1b[?25l\x1b[0m\x1b[?7l\x1b[0m\x1b[J[password]> ************                                                       \r\x1b[79C \r\x1b[24C\x1b[?7h\x1b[0m\x1b[?12l\x1b[?25h\x1b[?25l\x1b[?7l\x1b[?7h\x1b[0m\x1b[?12l\x1b[?25h\x1b[?25l\x1b[?7l\x1b[24D\x1b[0m\x1b[J[password]> ************                                                       \r\x1b[79C \r\x1b[0m\r\r\n\x1b[J\x1b[?7h\x1b[0m\x1b[?12l\x1b[?25h\x1b[?2004l"

----------------------------------------------------------------------
Ran 347 tests in 68.850s

FAILED (failures=7, errors=1)
```
I believe the major reason for the above failures, is related to this test case that I have added:
```
ERROR: test_initial_setup (api.JSONRPC.test_json_rpc_api.JsonRpcApiTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dau/workspace/python/github.com/neo/dautt/neo-python/neo/api/JSONRPC/test_json_rpc_api.py", line 62, in test_initial_setup
    self.assertTrue(GetBlockchain().GetBlock(0).Hash.To0xString(), '0x996e37358dc369912041f966f8c5d8d3a8255ba5dcbd3447f8a82b55db869099')
AttributeError: 'NoneType' object has no attribute 'Hash'
```
I tried to fix it but unsuccessfully. So I need help here. For some unknown reason the `Genesis block` is replaced by `None` when we run all the tests together, and therefore the final Blockchain is corrupted. (Strangely this behavior does not happen when we run the test one by one). If we can fix this one, most of the cases will be fix by themselves.

So how can we divide the remaining work?

I thought as follow, but please feel free to do as you wish. :smile:  
1) @ixje can perhaps fix the issue related to this command `python -m unittest discover neo`
2) @ixje  can review the test cases which depend only `WalletFixtureTestCase`. The one which depend on `BlockChainFixtureTestCase` have been already reviewed by @ixje  and threfore there is no need to review them again. I have labeled them as **okay**.
3) @jseagrave21 can update the test case depending only `BlockChainFixtureTestCase`. In particular  `test_json_rpc_api.py` and help @ixje  perhaps we the test case `test_claim_command.py`

**Comment by jseagrave21** [comment]
@dauTT Here is the migrated `test_json_rpc_api.py` including the new tests for `sendtoaddress` since @localhuman just merged that PR.

https://github.com/jseagrave21/neo-python/blob/DauTT-branch/neo/api/JSONRPC/test_json_rpc_api.py

Additionally, please note that the my "DauTT-branch" is completely up to date with your "latest2" branch, and it is also up-to-date with the latest CoZ Development branch. So, you should be able to merge my branch into your branch if you'd like.

**Comment by jseagrave21** [comment]
For the other failures in test_json_rpc_api.py when running `make test` there must be something wrong with the way that `discover neo` is being implemented.
For example, here is one of the failures:
```
FAIL: test_block_non_verbose (api.JSONRPC.test_json_rpc_api.JsonRpcApiTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/c/users/jseag/neo/neo-python-travis/neo/api/JSONRPC/test_json_rpc_api.py", line 241, in test_block_non_verbose
    self.assertEqual(block.Index, 2003)
AssertionError: 2002 != 2003
```
Yet, the block being queried in `test_json_rpc_api.py` is 2003:
```
def test_block_non_verbose(self):
        req = self._gen_rpc_req("getblock", params=[2003, 0])
        mock_req = mock_request(json.dumps(req).encode("utf-8"))
        res = json.loads(self.app.home(mock_req))
        self.assertIsNotNone(res['result'])

        # we should be able to instantiate a matching block with the result
        output = binascii.unhexlify(res['result'])
        block = Helper.AsSerializableWithType(output, 'neo.Core.Block.Block')
        self.assertEqual(block.Index, 2003)
        self.assertEqual(len(block.Transactions), 1)
```
So, the `discover neo` is off by one block.

**Comment by jseagrave21** [comment]
@dauTT, I am not sure about how to address this issue though perhaps it is related to #538 ?

**Comment by jseagrave21** [comment]
@dauTT I reran `make test` several times. I changed test_intial_setup to this:
```
def test_initial_setup(self):
        expected_blockhash = '0x996e37358dc369912041f966f8c5d8d3a8255ba5dcbd3447f8a82b55db869099'
        req = self._gen_rpc_req("getblockhash", params=[0])
        mock_req = mock_request(json.dumps(req).encode("utf-8"))
        res = json.loads(self.app.home(mock_req))
        self.assertEqual(expected_blockhash, res["result"])
```
Instead of the expected result, I consistently get `0xb3181718ef6167105b70920e4a8fbbd0a0a56aacf460d70e10ba6fa1668f1fef` which I cannot find when I query with `block` in np-prompt. So, it looks like block 0 is being replaced by a different block at some point during the `discover neo` test.

**Comment by ixje** [comment]
@dauTT 
1) I solved the `unittest discovery` difference. You'll have to merge this PR: https://github.com/CityOfZion/neo-python/pull/616

2) Having this call is good https://github.com/dauTT/neo-python/blob/latest2-move-unittest-to-privnet-478/neo/Settings.py#L232-L234
But we do not want to copy the settings from the privnet protocol file. We should create a new one and check that in (e.g. `protocol.unittest-net.json`). It can be a copy of the privnet one (content wise), but they cannot be the same file. Reason being is that people should be free to change their privatenet protocol to whatever matches their needs, however they might touch e.g. the `validators` or `magic` and that will mess up the test cases. 

3) this test will still fail
https://github.com/CityOfZion/neo-python/blob/951de3a902a398ce6b178481e837ce96014affbd/neo/Implementations/Blockchains/LevelDB/tests/test_initial_db.py#L45-L56

You'll have to update the `setUpClass()`to add:
```
def setUpClass(self):
        settings.setup_fixturenet()
        Blockchain.DeregisterBlockchain()
```
and update the test to the following:
```
    def test_add_header(self):
        hexdata = binascii.unhexlify(self.block_one_raw)
        block_one = Helper.AsSerializableWithType(hexdata, 'neo.Core.Block.Block')

        if settings.MAGIC == 56753:
            self.assertEqual(self._blockchain.CurrentHeaderHash, b'996e37358dc369912041f966f8c5d8d3a8255ba5dcbd3447f8a82b55db869099')
        else:
            self.assertEqual(self._blockchain.CurrentHeaderHash, b'd42561e3d30e15be6400b6df2f328e02d2bf6354c41dce433bc57687c82144bf')

        self.assertEqual(self._blockchain.HeaderHeight, 0)

        self._blockchain.AddBlock(block_one)
        self.assertEqual(self._blockchain.HeaderHeight, 1)
```
that will fix the errors and actually make the test do what the testname suggests 😂

**Comment by dauTT** [comment]
Thanks @ixje / @jseagrave21 , great team work!!! :smile: 

I have created a clean branch with all the PR (611, 616) merged together along with the update file `test_json_rpc_api` from @jseagrave21 branch. Here is the [link](https://github.com/dauTT/neo-python/tree/merge-pr616-pr611-into-latest2-move-unittest-to-privnet-478)

**Fixtures to upload to AWS:**

[notif_fixtures_v8.tar.gz](https://github.com/CityOfZion/neo-python/files/2394745/notif_fixtures_v8.tar.gz)
[fixtures_v8.tar.gz](https://github.com/CityOfZion/neo-python/files/2394746/fixtures_v8.tar.gz)

**Tested the code by running:**
``` 
make lint
make test
```
**Everything pass except this one:**
```
======================================================================
FAIL: test_prompt_open_wallet (bin.test_prompt.PromptTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dau/workspace/python/github.com/neo/dautt/neo-python/neo/bin/test_prompt.py", line 26, in test_prompt_open_wallet
    self.assertIn(checktext, text)
AssertionError: 'Opened' not found in "open wallet fixtures/testwallet.db3\r\ntestpassword\r\n\x1b[32m[I 180919 00:31:55 LevelDBBlockchain:113]\x1b[39m Created Blockchain DB at /home/dau/.neopython/Chains/SC234 \r\n\x1b[32m[I 180919 00:31:58 NotificationDB:73]\x1b[39m Created Notification DB At /home/dau/.neopython/Chains/Test_Notif \r\n\x1b[0m\x1b[?7h\x1b[0;38;5;4mNEO\x1b[0;38;5;10m cli. Type \x1b[0;38;5;197m'help' \x1b[0;38;5;10mto get started\x1b[0m\r\r\n\x1b[0m\r\n\r\n\x1b[?1l\x1b[6n\x1b[?2004h\x1b[?25l\x1b[0m\x1b[?7l\x1b[0m\x1b[Jneo>  \x1b[0m\r\r\n\r\r\n\r\r\n\r\r\n\r\r\n\r\r\n\r\r\n                                                                               \r\x1b[79C \r\x1b[7A\x1b[5C\x1b[?7h\x1b[0m\x1b[?12l\x1b[?25h\x1b[?25l\x1b[?7l\x1b[5Dneo> open wallet fixtures/testwallet.db3                                       \r\x1b[79C \r\x1b[40C\x1b[?7h\x1b[0m\x1b[?12l\x1b[?25h\x1b[?25l\x1b[?7l\x1b[40D\x1b[0m\x1b[Jneo> open wallet fixtures/testwallet.db3                                       \r\x1b[79C \r\x1b[0m\r\r\n\x1b[J\x1b[?7h\x1b[0m\x1b[?12l\x1b[?25h\x1b[?2004l\x1b[?1l\x1b[?2004h\x1b[?25l\x1b[0m\x1b[?7l\x1b[0m\x1b[J[password]> ************                                                       \r\x1b[79C \r\x1b[24C\x1b[?7h\x1b[0m\x1b[?12l\x1b[?25h\x1b[?25l\x1b[?7l\x1b[?7h\x1b[0m\x1b[?12l\x1b[?25h\x1b[?25l\x1b[?7l\x1b[24D\x1b[0m\x1b[J[password]> ************                                                       \r\x1b[79C \r\x1b[0m\r\r\n\x1b[J\x1b[?7h\x1b[0m\x1b[?12l\x1b[?25h\x1b[?2004l"

----------------------------------------------------------------------
Ran 352 tests in 71.690s

FAILED (failures=1)
```
This was also failing before we did any changes. It has nothing to do with `BlockChainFixtureTestCase` nor `WalletFixtureTestCase`. I am not sure if this is just happening on my local machine or not.
 
@jseagrave21 , I believe you know very well this test. Can you please have a look at it? Thanks. :smile:

**Comment by jseagrave21** [comment]
@dauTT okay let me take a look

**Comment by jseagrave21** [comment]
@dauTT since we didn't change the test and testwallet.db3 is still accessible, I suspect it is just the time out. I added 5 sec to the test here. Feel free to add time if you need to.
https://github.com/jseagrave21/neo-python/blob/DauTT-branch/neo/bin/test_prompt.py

**Comment by dauTT** [comment]
Yes, your right, increasing the time out help. So all the test cases are fine. :+1:
